### ISSUE TITLE
keyed_hash and derive_key support

### ISSUE THREAD

**Issue by tv42** [zeebo/blake3#4]
Hi. Great work, this is really neat to get the true ASM speed going.

I'd love to migrate some of my projects to blake3, but I'm a heavy user of keyed hashes. Any chances of support for the keyed hash and key derivation modes?
See e.g. https://github.com/BLAKE3-team/BLAKE3#the-blake3-crate

(And then there's XOF, but I don't personally need that right now...)

~~API-wise I'd expect `NewWithOptions` / `NewXOFWithOptions` from https://github.com/golang/go/issues/32447#issuecomment-597305745 would seem like the way to go (with the Salt and Personalization fields).~~ (not convinced about that anymore)

**Comment by zeebo** [comment]
Those things shouldn't be too hard to implement. I think most of the plumbing for them is already there, I just didn't wire it up because I was grappling with implementing anything working in the first place. I'm obviously not against adding them.

That said, I don't know if I'll have much time to work on it in the near future.

**Comment by tv42** [comment]
Here's an initial commit for NewKeyed. Pure Go only at this stage. Most of the changes revolve around plumbing for the right key (instead of hardcoding iv) and flags (instead of hardcoding 0). What do you think?

https://github.com/tv42/blake3/commit/23a2906c4ff2fd5aa7294777818fb94cb58c76cf



As for XOF, I've been thinking blake3.Hasher would get an extra method `XOF() io.Reader`, but I'm not clear on what that means for the hash state; as in, can you call XOF more than once? Is the Hasher still valid to use after XOF()? Etc.

https://godoc.org/golang.org/x/crypto/blake2b#XOF is a very different API, but afaik blake3 doesn't need a "special mode" to be XOF, it's basically a replacement for the `Sum` method.

**Comment by zeebo** [comment]
So far that looks on track with what I'd expect. Awesome.

As far as XOF stuff is concerned, I think the Rust API is such that you can create as many copies of the state as you want, and they're all independent.

I think this should be achievable in basically the same way as how Write is allowed to happen after a Sum call. IIRC the XOF output is essentially a block counter around the finalized sum plus some flags.

**Comment by zeebo** [comment]
You have managed to successfully nerd snipe me into getting avx2 support added for keyed hashes. Surprisingly, that was the easiest part: just a 6 loc diff ([here](https://github.com/zeebo/blake3/compare/keyed?expand=1#diff-36aa054ac726978e152e45f6606f00a8) and [here](https://github.com/zeebo/blake3/compare/keyed?expand=1#diff-109b497848e9ba4e05b23b1eebb882a3)) . You had already done most of the work :smile: 

So then I just ended up spending way too long polishing, increasing and improving test coverage, refactoring, and adding XOF support. The work is in the keyed branch. I'll probably be spending a bit more time polishing, and I'd love to hear any feedback you may have.

**Comment by tv42** [comment]
I ported my little project to proper keyed hashes and XOF (instead of using blake2, or faking them with feeding the key to the hash as prefix and chacha20 streams). The API shape is reasonable and everything works for me. Note that this is not a statement for the hash results being correct because this is the only implementation in existence using blake3 for this. But I see no crashes, reproducible results, etc, in light usage.

Next up, if you don't beat me to it, I'll work on key derivation. It should be easy after the IV plumbing. For that, one question: do you prefer

```
type constantString string

func DeriveKey(context constantString) *blake3.Hasher
func DeriveKeySized(context constantString, size int) (*blake3.Hasher, error)
```

or just with regular `string`? The `constantString` trick forces callers to use a constant there, which is very much what the blake3 spec says you really really should do -- but it prevents using any kind of helper functions wrapping that call that would take the context as an argument.

**Comment by tv42** [comment]
First stab at DeriveKey at https://github.com/tv42/blake3/commit/6812f66d1debb088819a11023c68371f6d65064e

The unit test came out looking quite different, the way key derivation works it's not just a single `hasher` invocation.

Let me know what you think.

**Comment by zeebo** [comment]
Is there a reason why you decided to have `DeriveKey` both accept the material and return a `Hasher`? I would expect it to either

1. Accept the material and return an XOF
2. Accept the material and an output buffer ([rust library](https://github.com/BLAKE3-team/BLAKE3/blob/master/src/lib.rs#L767))
3. Not accept the material and return a `Hasher` ([rust library](https://github.com/BLAKE3-team/BLAKE3/blob/master/src/lib.rs#L869))

I think if it's named `DeriveKey{,Sized}` it should probably be of form 1 or 2, but if it's called something like `NewKeyDerivation{,Sized}` it should be of form 3.

---

Secondary thought: since we have the `XOF` option, do we want to bother with with the all the `Sized` variants? You can just call `XOF().Read(buf)` and pick buf to be the right size.

**Comment by tv42** [comment]
Sorry for the delay, I missed the notification for this.

I put in the `material` argument because I expected it to be very common for callers to be holding a '[]byte` of the key material, and I copied the calling style I saw elsewhere (pdf/rust reference implementation, I believe).

Option 1 seems good to me. The code was written before the XOF type existed, so I didn't really even think much about that.

I have personal use for unknown-length XOF output from key derivation (length needed depends on data seen; basically CSPRNG until values fulfill criteria), so option 2 is not interesting to me.

The biggest downside with returning `XOF` from `DeriveKey` is that it makes it clumsier for people who want a fixed-size output and are accustomed to `hash.Hash`. But I'd still say let's do that. That also removes the confusing `XxxSized().XOF()` possibility

So just

```
func DeriveKey(context string, material []byte) *XOF
```

and if we wanted to support option 3, also

```
func NewDeriveKey(context string) *Hasher
```

**Comment by tv42** [comment]
Of course if option 3 is available for the advanced use cases, option 2 is tempting with its simpler API for basic use. I'm switching my opinion to favor cloning the rust API:

```
func DeriveKey(context string, material []byte, out []byte)
func NewDeriveKey(context string) *Hasher
```

**Comment by tv42** [comment]
Here's a commit with the new calling style: https://github.com/tv42/blake3/commit/e86e7e9532fa4bd8f18bd23d009ac62dbcc6030b

Let me know if you prefer pull requests. Github ain't great with rebasing commits in one pull request so I don't like doing them until I have something acceptable, and multiple PRs for such a little thing just seem too noisy.

**Comment by zeebo** [comment]
Cool. I agree with that API.

I've just been fetching from your repo and cherry-picking and it has been working fine.

I thinking about renaming the `XOF` method and type to be `Output`, or `Digest` or something. Any thoughts?

Lastly, do you have any thoughts on removing the `Sized` constructor variants in favor of just having people use those methods?

---

I think this is really close. Just these last questions about some minor API details. Thanks for sticking with me through it.

**Comment by tv42** [comment]
> I thinking about renaming the XOF method and type to be Output, or Digest or something. Any thoughts?

That's a tough one. The crypto community seems to be happily saying XOF, and there's https://godoc.org/golang.org/x/crypto/blake2b#XOF

`Digest` could be taken to mean the `[]byte` containing the hash sum. "md5 digest" has long meant the output of the hash. And `Output` is one those words that sort of doesn't mean anything. I'm personally ok with `XOF`, even if I normally hate TLAs -- it's just the ecosystem's fault that we don't have a better word. You could name it `Reader`.

> Lastly, do you have any thoughts on removing the Sized constructor variants in favor of just having people use those methods?

I removed/replaced `DeriveKeySized` in my latest commit. If you mean `NewSized`/`NewKeyedSized`, well, those are sort of dictated by the `hash.Hash` interface.

**Comment by zeebo** [comment]
>  Digest could be taken to mean the []byte containing the hash sum. "md5 digest" has long meant the output of the hash. And Output is one those words that sort of doesn't mean anything. I'm personally ok with XOF, even if I normally hate TLAs -- it's just the ecosystem's fault that we don't have a better word. You could name it Reader (the type, not the method).

I mostly agree. Output is definitely one of those meaningless words. But the more I think about it, the more I like `Digest`. In a sense, it IS the output `[]byte`, just a lazily generated, seekable one. Returning some named type representing that seems reasonable.

> I removed/replaced DeriveKeySized in my latest commit. If you mean NewSized/NewKeyedSized, well, those are sort of dictated by the hash.Hash interface.

It seems likely to me that if you care about the output size, you can use the concrete hash.Hash. Maybe not. Maybe we just add a `SetSize` method on the concrete type so that you can just do something like

```go
h := blake3.New()
h.SetSize(16)
doSomethingGeneric(h)
```

Then at least the API surface is O(1) instead of O(number of constructors).

**Comment by tv42** [comment]
`Digest` is fine if you want it.

`Hasher.SetSize` doesn't look like most good Go :-(

Maybe use the Option trick?

```
h1, err := blake3.New()
h2, err := blake3.New(blake3.Sized(42))
h3, err := blake3.New(blake3.Keyed(key))
h4, err := blake3.New(blake3.Derived("example.com foo bar"))
h5, err := blake3.New(
        blake3.Derived("example.com foo bar"),
        blake3.Sized(42),
        )
```

Err is there because both Sized and Keyed can complain about bad input. I really enjoy the fact that `DeriveKey`/`NewDeriveKey` in https://github.com/zeebo/blake3/issues/4#issuecomment-613674797 doesn't have errors, so mixing those with Sized and Keyed bums me out. Actually, I hate that so much that I now don't like this API idea.

**Comment by tv42** [comment]
This really doesn't seem like a too large API, to me. Having some Sized variants and not all is a bit weird though, but it's `hash.Hash` and hashes that can't XOF causing that problem...

```
func New() *Hasher {
func NewSized(size int) (*Hasher, error) {
func NewKeyed(key []byte) (*Hasher, error) {
func NewKeyedSized(key []byte, size int) (*Hasher, error) {
func DeriveKey(context string, material []byte, out []byte) {
func NewDeriveKey(context string) *Hasher {
```

With `Hasher.SetSize`, that would drop down to

```
func New() *Hasher {
func NewKeyed(key []byte) (*Hasher, error) {
func DeriveKey(context string, material []byte, out []byte) {
func NewDeriveKey(context string) *Hasher {
```

which doesn't seem like a large enough win.

**Comment by zeebo** [comment]
Hmm. Yeah, those are all great points. I'm going to think about it for a day or two. Feel free to drop in any other ideas you may have.

**Comment by zeebo** [comment]
Ok sat on it for far too long. I'm just going to decide to go with whatever interface is minimal, even if it makes certain operations harder. It's easy to expand in the future, and I'd like to keep the library as stable as possible.

Thus, I'm just going to remove the Sized constructors and SetSize, and if you need a different size of output, use XOF.

**Comment by zeebo** [comment]
Fixed by #6

**Comment by tv42** [comment]
I switched my stuff to v0.0.3 and I'm very happy with the API.
